presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,442 @@
+
+    // The number of positions (rows) buffered so far
+    protected int bufferedPositionCount;
+
+    // Whether the positions array was mapped to mappedPositions
+    protected boolean isPositionsMapped;
+
+    // The positions (row numbers) of the incoming Block to be copied to this buffer.
+    // All top level BlockEncodingBuffers for the same partition share the same positions array.
+    private int[] positions;
+
+    // The mapped positions of the incoming Dictionary or RLE Block
+    @Nullable
+    private int[] mappedPositions;
+
+    @Nullable
+    private byte[] nullsBuffer;
+
+    // The address offset in the nullsBuffer if new values are to be added.
+    private int nullsBufferIndex;
+
+    // For each batch we put the nulls values up to a multiple of 8 into nullsBuffer. The rest is kept in remainingNullsFromLastBatch.
+    private final boolean[] remainingNullsFromLastBatch = new boolean[BITS_IN_BYTE];
+
+    // Number of null values not put into nullsBuffer from last batch
+    private int remainingNullsCount;
+
+    // Whether nullsBuffer contains nulls. It is possible that the nullsBuffer contains all non-nulls.
+    private boolean nullsBufferContainsNull;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(DecodedBlockNode decodedBlockNode, int initialPositionCount)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        Object decodedBlock = decodedBlockNode.getDecodedBlock();
+
+        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
+        if (decodedBlock instanceof DictionaryBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+
+        if (decodedBlock instanceof RunLengthEncodedBlock) {
+            decodedBlockNode = decodedBlockNode.getChildren().get(0);
+            decodedBlock = decodedBlockNode.getDecodedBlock();
+        }
+
+        if (decodedBlock instanceof LongArrayBlock) {
+            return new LongArrayBlockEncodingBuffers(initialPositionCount);
+        }
+
+        throw new IllegalArgumentException("Unsupported encoding: " + decodedBlock.getClass().getSimpleName());
+    }
+
+    public BlockEncodingBuffers(int initialPositionCount)
+    {
+        this.initialPositionCount = initialPositionCount;
+    }
+
+    public abstract void resetBuffers();
+
+    protected abstract void prepareBuffers();
+
+    public void setNextBatch(int positionsOffset, int batchSize)
+    {
+        this.positionsOffset = positionsOffset;
+        this.batchSize = batchSize;
+    }
+
+    public abstract void copyValues();
+
+    public abstract void serializeTo(SliceOutput sliceOutput);
+
+    public abstract long getSizeInBytes();
+
+    public abstract long getRetainedSizeInBytes();
+
+    public abstract long getSerializedSizeInBytes();
+
+    public abstract void accumulateRowSizes(int[] rowSizes);
```

Comment:
Hi Masha, I think rowSize here is more explanatory. Position size could be for a single column, or could be multiple columns, but row size is straightforward. I changed the appendRows function to appendData though. @tdcmeehan what do you think?

Reply:
The comment says `Adds serialized row sizes to serializedRowSizes array.` . So why not just name the method `accumulateSerializedRowSizes` and use `SerializedRowSize` instead of `RowSize` in other places as well? ;)